lib/horde/supervisor_impl.ex

```diff
@@ -345,11 +327,20 @@ defmodule Horde.SupervisorImpl do
 
     MapSet.difference(old_process_ids, new_process_ids)
     |> Enum.map(fn id ->
-      :ok =
-        DynamicSupervisor.terminate_child(
+      pid =
+        Horde.DynamicSupervisor.find_pid(
           processes_supervisor_name(new_state.name),
-          get_child_pid(new_state, id)
+          id
         )
+
+      if pid do
```

Comment:
Ok, so here we are checking whether the pid exists, but on line 92 we do not. Is there a race condition here that I'm not aware of?

My thinking is: the DynamicSupervisor will *always* have a pid for every process under its supervision (if it could not start a child process after X retries within Y minutes, the supervisor itself will die).

Actually come to think of it, the following could happen:
1. we call `Horde.DynamicSupervisor.find_pid`
2. the process belonging to the pid that is returned dies
3. Horde.DynamicSupervisor restarts the process (now has a new pid, and doesn't know about our pid anymore)
4. We call Horde.DynamicSupervisor.terminate_child

This indicates to me that we need to bundle these two calls into one: a terminate_child that accepts an `id` instead of a `pid`.

Reply:
And if we have this, then we can remove all mention of child pids from this module (should save some bookkeeping / code).

Reply:
So with the old code, we had an implied concept of local ownership(Registry) and global ownership(CRDT). With Process Canary removed we lost the concept of local ownership so what can happen here is when the Supervisor tries to kill it's not owned children all it has left is the global data which is a race i.e. often what would happen is that it would try to remove a pid that didn't belong to this supervisor. So what I had to do was give the local supervisor the ability to map child_ids to pids. I'm checking here just to make sure that the pid exists and is not nil maybe unnecessarily. The reason I don't check in 98 is that method affects global data whereas this one is pure local data.

Reply:
Ah I think I see your point this can't be two different calls otherwise the race will occur.